### PR TITLE
feat:refactor smart contract

### PR DESCRIPTION
**Pull request by vuvth**
- Remove builds folder (because we are using monorepo so we don't need it anymore) 
- Remove duplicate code in smart contract(Unirep.sol) 
- Update script to build veirifier contract more clean(check contracts/scripts/buildVerifier.sh) 
- Create contracts/interfaces
- Create contracts/libraries
- Create contracts/types
- Create libaries/znSNARKHelper.sol (check snark field)

### CODE REVIEW

**Review by vivianjeng — APPROVED** [review]
just minor changes.
I think it looks good

File: packages/contracts/contracts/interfaces/IUnirep.sol [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        uint256 _outputBlindedUserState,
+        uint256 _outputBlindedHashChain,
+        uint256[8] _proof
+    );
+
+    event IndexedUserStateTransitionProof(
+        uint256 indexed _proofIndex,
+        UserTransitionProof _proof,
+        uint256[] _proofIndexRecords
+    );
```
[the remark this answers is not in the document]
Comment (vuvth) [reply packages/contracts/contracts/interfaces/IUnirep.sol]: fixed ❤️

File: packages/contracts/contracts/interfaces/IUnirep.sol [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    enum AttestationEvent {
+        SendAttestation,
+        Airdrop,
+        SpendReputation
+    }
+
+      // Events
+    event Sequencer(
+        uint256 indexed epoch, 
+        Event userEvent
```
Comment (vuvth) [inline packages/contracts/contracts/interfaces/IUnirep.sol:22]: I rename `_event` to `userEvent`. @vivianjeng

File: packages/contracts/contracts/interfaces/IUnirep.sol [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        uint256 indexed epoch,
+        uint256 indexed hashedLeaf,
+        uint256 proofIndex
+    );
+
+    event AttestationSubmitted(
+        uint256 indexed epoch,
+        uint256 indexed epochKey,
+        address indexed attester,
+        AttestationEvent attestationEvent,
```
Comment (vuvth) [inline packages/contracts/contracts/interfaces/IUnirep.sol:45]: I rename `_event` to `attestationEvent`. @vivianjeng

**Review by vivianjeng — APPROVED** [review]
LGTM

**Pull request merged**